### PR TITLE
Feature(Reservations): UI-31879: Reservations

### PR DESCRIPTION
**Pull request by sreekarbvibm**
Tested on Prod Sydney region with Beta maturity. Attached evidence in https://jiracloud.swg.usma.ibm.com:8443/browse/UI-26980

### CODE REVIEW

**Review by astha-jain — COMMENTED** [review]
instance reservation changes?

**Comment by sreekarbvibm** [comment]
> instance reservation changes?

Hi Astha, I will raise a separate PR for that. This is for reservation feature CRUD+Activate APIs

File: website/docs/d/is_reservation.html.markdown [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+```terraform
+provider "ibm" {
+  region = "eu-gb"
+}
+```
+
+## Example usage
+// Example to retrieve the reservation information by using reservation name.
+
+```terraform
```
Comment (SunithaGudisagarIBM1) [inline website/docs/d/is_reservation.html.markdown:26]: please run terraform fmt and paste here

File: website/docs/d/is_reservation.html.markdown [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+page_title: "IBM : reservation"
+description: |-
+  Manages IBM Cloud reservations.
+---
+
+# ibm_is_reservation
+Retrieve information of an existing reservation as a read only data source.
+
+**Note:** 
+VPC infrastructure services are a regional specific based endpoint, by default targets to `us-south`. Please make sure to target right region in the provider block as shown in the `provider.tf` file, if VPC service is created in region other than `us-south`.
```
Comment (SunithaGudisagarIBM1) [inline website/docs/d/is_reservation.html.markdown:13]: mention the link when its going to upstream. refer other resources

File: website/docs/d/is_reservation.html.markdown [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+}
+data "ibm_is_reservation" "ds_res" {
+	name = ibm_is_reservation.res.name
+}
+
+```
+// Example to retrieve the reservation information by using reservation ID.
+
+```terraform
+resource "ibm_is_reservation" "res" {
```
Comment (SunithaGudisagarIBM1) [inline website/docs/d/is_reservation.html.markdown:49]: please run terraform fmt and paste here

File: website/docs/d/is_reservation.html.markdown [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+			  zone = "us-east-3"
+			  name = "reservation-name"
+		}
+		data "ibm_is_reservation" "ds_res" {
+		    identifier = ibm_is_reservation.res.id
+		}
+
+```
+
+## Argument reference
```
Comment (SunithaGudisagarIBM1) [inline website/docs/d/is_reservation.html.markdown:69]: check in preview tool for the alignment

File: website/docs/d/is_reservations.html.markdown [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+```terraform
+provider "ibm" {
+  region = "eu-gb"
+}
+```
+
+## Example usage
+
+```terraform
+
```
Comment (SunithaGudisagarIBM1) [inline website/docs/d/is_reservations.html.markdown:26]: please run terraform fmt and paste here

File: website/docs/d/is_reservations.html.markdown [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    - `href` - (String) The URL for this resource group.
+    - `id` - (String) The unique identifier for this resource group.
+    - `name` - (String) The name for this resource group.
+  - `resource_type` - (String) The resource type.
+
+    ->**resource_type** 
+      </br>&#x2022; reservation
+  - `status` - (String) The status of the reservation.
+
+    ->**status**
```
Comment (SunithaGudisagarIBM1) [inline website/docs/d/is_reservations.html.markdown:134]: Can you check other doc and see that we are mentioning allowed values this way.. i feel it makes the doc very lengthy..

File: website/docs/d/is_reservations.html.markdown [hunk 7]
```diff
@@ -0,0 +1,151 @@
+---
+subcategory: "VPC infrastructure"
+layout: "ibm"
+page_title: "IBM : Reservations"
+description: |-
+  Manages IBM Cloud infrastructure reservations.
+---
+
+# ibm_is_reservations
+Retrieve information of an existing reservations as a read only data source.
```
Comment (SunithaGudisagarIBM1) [inline website/docs/d/is_reservations.html.markdown:10]: link here please

File: website/docs/r/is_reservation.html.markdown [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+page_title: "IBM : reservation"
+description: |-
+  Manages IBM reservation.
+---
+
+# ibm_is_reservation
+Create, update, or delete a reservation.
+
+**Note:** 
+VPC infrastructure services are a regional specific based endpoint, by default targets to `us-south`. Please make sure to target right region in the provider block as shown in the `provider.tf` file, if VPC service is created in region other than `us-south`.
```
Comment (SunithaGudisagarIBM1) [inline website/docs/r/is_reservation.html.markdown:14]: link please

**Review by SunithaGudisagarIBM1 — CHANGES_REQUESTED** [review]
Hi Sreekar,

Please find my review comments in documentation. 

Thanks 
Sunitha

File: website/docs/r/is_reservation.html.markdown (hunk 8 above)
Comment (sreekarbvibm) [reply website/docs/r/is_reservation.html.markdown:14]: There is no link yet, I will update once links are available

File: website/docs/d/is_reservations.html.markdown (hunk 7 above)
Comment (sreekarbvibm) [reply website/docs/d/is_reservations.html.markdown:10]: There is no link yet, I will update once links are available

File: website/docs/d/is_reservations.html.markdown (hunk 6 above)
Comment (sreekarbvibm) [reply website/docs/d/is_reservations.html.markdown:134]: Yes, I referred to other is_* doc(s) and created this.

File: website/docs/d/is_reservations.html.markdown (hunk 5 above)
Comment (sreekarbvibm) [reply website/docs/d/is_reservations.html.markdown:26]: done

File: website/docs/d/is_reservation.html.markdown (hunk 4 above)
Comment (sreekarbvibm) [reply website/docs/d/is_reservation.html.markdown:69]: done

File: website/docs/d/is_reservation.html.markdown (hunk 3 above)
Comment (sreekarbvibm) [reply website/docs/d/is_reservation.html.markdown:49]: done

File: website/docs/d/is_reservation.html.markdown (hunk 2 above)
Comment (sreekarbvibm) [reply website/docs/d/is_reservation.html.markdown:13]: There is no link yet, I will update once links are available

File: website/docs/d/is_reservation.html.markdown (hunk 1 above)
Comment (sreekarbvibm) [reply website/docs/d/is_reservation.html.markdown:26]: Done

File: ibm/service/vpc/data_source_ibm_is_reservation.go [hunk 9]
```diff
@@ -0,0 +1,430 @@
...
+		getReservationOptions.SetID(id)
+		reservationInfo, response, err := sess.GetReservationWithContext(context, getReservationOptions)
+		if err != nil {
+			log.Printf("[DEBUG] GetReservationWithContext failed %s\n%s", err, response)
+			return diag.FromErr(fmt.Errorf("[ERROR] GetReservationWithContext failed %s\n%s", err, response))
+		}
+		reservation = reservationInfo
+
+	} else if v, ok := d.GetOk("name"); ok {
+
```
Comment (sumitkumartiwari) [inline ibm/service/vpc/data_source_ibm_is_reservation.go:240]: Do we have validator to validate id or name as mutually exclusive for filter

File: ibm/service/vpc/data_source_ibm_is_reservation_test.go [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+			  }
+			  profile {
+				name = "ba2-2x8"
+				resource_type = "instance_profile"
+			  }
+			  zone = "us-east-3"
+			  name = "reservation-name"
+		}
+		data "ibm_is_reservation" "ds_res" {
+		    name = ibm_is_reservation.res.id
```
Comment (sumitkumartiwari) [inline ibm/service/vpc/data_source_ibm_is_reservation_test.go:56]: name assigned by ID value.

File: ibm/service/vpc/data_source_ibm_is_reservations.go [hunk 11]
```diff
@@ -0,0 +1,452 @@
...
+	sess, err := vpcClient(meta)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	zoneName := d.Get("zone_name").(string)
+	resourceGroupId := d.Get("resource_group").(string)
+
+	start := ""
+	matchReservations := []vpcv1.Reservation{}
```
Comment (sumitkumartiwari) [inline ibm/service/vpc/data_source_ibm_is_reservations.go:245]: can we put more meaningful variable name matchReservations ? like reservations or listReservations ?

File: ibm/service/vpc/data_source_ibm_is_reservation_test.go [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+					resource.TestCheckResourceAttrSet("data.ibm_is_reservation.ds_res", "created_at"),
+					resource.TestCheckResourceAttrSet("data.ibm_is_reservation.ds_res", "crn"),
+					resource.TestCheckResourceAttrSet("data.ibm_is_reservation.ds_res", "href"),
+					resource.TestCheckResourceAttrSet("data.ibm_is_reservation.ds_res", "lifecycle_state"),
+					resource.TestCheckResourceAttrSet("data.ibm_is_reservation.ds_res", "name"),
+					resource.TestCheckResourceAttrSet("data.ibm_is_reservation.ds_res", "profile.#"),
+					resource.TestCheckResourceAttrSet("data.ibm_is_reservation.ds_res", "resource_group.#"),
+					resource.TestCheckResourceAttrSet("data.ibm_is_reservation.ds_res", "resource_type"),
+					resource.TestCheckResourceAttrSet("data.ibm_is_reservation.ds_res", "zone"),
+				),
```
Comment (sumitkumartiwari) [inline ibm/service/vpc/data_source_ibm_is_reservation_test.go:33]: add two test cases as filter for each by ID and Name

File: website/docs/r/is_instance.html.markdown [hunk 13]
```diff
@@ -127,6 +127,86 @@ resource "ibm_is_instance" "example1" {
...
+  zone            = "us-south-1"
+  ipv4_cidr_block = "10.240.0.0/24"
+}
+
+resource "ibm_is_ssh_key" "example" {
+  name       = "example-ssh"
+  public_key = "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQCKVmnMOlHKcZK8tpt3MP1lqOLAcqcJzhsvJcjscgVERRN7/9484SOBJ3HSKxxNG5JN8owAjy5f9yYwcUg+JaUVuytn5Pv3aeYROHGGg+5G346xaq3DAwX6Y5ykr2fvjObgncQBnuU5KHWCECO/4h8uWuwh/kfniXPVjFToc+gnkqA+3RKpAecZhFXwfalQ9mMuYGFxn+fwn8cYEApsJbsEmb0iJwPiZ5hjFC8wREuiTlhPHDgkBLOiycd20op2nXzDbHfCHInquEe/gYxEitALONxm0swBOwJZwlTDOB7C6y2dzlrtxr1L59m7pCkWI4EtTRLvleehBoj3u7jB4usR"
+}
+
+resource "ibm_is_reservation" "example" {
```
Comment (sumitkumartiwari) [inline website/docs/r/is_instance.html.markdown:149]: From here only required

File: website/docs/r/is_reservation_activate.html.markdown [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+Provide support to activate a reservation. This resource activates a reservation, resulting in its status becoming `active`.
+
+**Note:** 
+VPC infrastructure services are a regional specific based endpoint, by default targets to `us-south`. Please make sure to target right region in the provider block as shown in the `provider.tf` file, if VPC service is created in region other than `us-south`.
+
+**provider.tf**
+
+```terraform
+provider "ibm" {
+  region = "eu-gb"
```
Comment (sumitkumartiwari) [inline website/docs/r/is_reservation_activate.html.markdown:21]: check this

**Comment by sreekarbvibm** [comment]
output_inst_tmpl_create_with_res
code_inst_tmpl_create_with_res
output_get_inst_with_res
code_get_inst_with_res
output_inst_update_detach_res
code_inst_update_detach_res
output_inst_create_with_res_2
output_inst_create_with_res_1
code_inst_create_with_res
output_update_res
code_update_res
output_get_res
code_get_res
output_list_res_2
output_list_res_1
code_list_res
output_res_activate
code_res_activate
output_res_create
code_res_create

**Comment by sreekarbvibm** [comment]
> > instance reservation changes?
> 
> Hi Astha, I will raise a separate PR for that. This is for reservation feature CRUD+Activate APIs

Updated same PR

File: ibm/service/vpc/data_source_ibm_is_reservation.go (hunk 9 above)
Comment (sreekarbvibm) [reply ibm/service/vpc/data_source_ibm_is_reservation.go:240]: We get below error on terraform apply.
identifier": only one of `identifier,name` can be specified, but `identifier,name` were specified.

Removed else, kept it as separate if statement.

File: ibm/service/vpc/data_source_ibm_is_reservations.go (hunk 11 above)
Comment (sreekarbvibm) [reply ibm/service/vpc/data_source_ibm_is_reservations.go:245]: renamed to reservations

File: website/docs/r/is_instance_template.html.markdown [hunk 15]
```diff
@@ -269,6 +269,16 @@ Review the argument references that you can specify for your resource.
...
+- `reservation_affinity` - (Optional, List) The reservation affinity for the instance
+  Nested scheme for `reservation_affinity`:
+  - `policy` - (Optional, String) The reservation affinity policy to use for this virtual server instance.
+
+     ->**policy** 
+			&#x2022; disabled: Reservations will not be used
+      </br>&#x2022; manual: Reservations in pool will be available for use
+  - `pool` - (Optional, String) The pool of reservations available for use by this virtual server instance. Specified reservations must have a status of active, and have the same profile and zone as this virtual server instance. The pool must be empty if policy is disabled, and must not be empty if policy is manual.
+    Nested scheme for `pool`:
+    - `id` - The unique identifier for this reservation
```
Comment (astha-jain) [inline website/docs/r/is_instance_template.html.markdown:281]: Same here

Comment (sreekarbvibm) [reply website/docs/r/is_instance_template.html.markdown:281]: Yes, but maxItems is 1

**Review by astha-jain — APPROVED** [review]

**Comment by sreekarbvibm** [comment]
Upstream PR is created - https://github.com/IBM-Cloud/terraform-provider-ibm/pull/5089

**Pull request closed without merging**